src/modules/job-manager/job.c

```diff
@@ -161,21 +163,38 @@ struct job *job_create_from_eventlog (flux_jobid_t id,
         return NULL;
     job->id = id;
 
-    if (!(job->jobspec_redacted = json_loads (jobspec, 0, NULL)))
+    if (!(job->jobspec_redacted = json_loads (jobspec, 0, NULL))) {
+        errprintf (error, "failed to decode jobspec");
         goto inval;
+    }
     jpath_del (job->jobspec_redacted, "attributes.system.environment");
 
-    if (!(a = eventlog_decode (eventlog)))
+    if (!(a = eventlog_decode (eventlog))) {
+        errprintf (error, "failed to decode eventlog");
         goto error;
+    }
 
     json_array_foreach (a, index, event) {
-        if (event_job_update (job, event) < 0)
+        const char *name = "unknown";
+        (void)eventlog_entry_parse (event, NULL, &name, NULL);
+
+        if (index == 0 && !streq (name, "submit")) {
+            errprintf (error, "first event is %s not submit", name);
+            goto inval;
+        }
+        if (event_job_update (job, event) < 0) {
+            errprintf (error, "could not apply %s", name);
             goto error;
+        }
         job->eventlog_seq++;
     }
 
-    if (job->state == FLUX_JOB_STATE_NEW)
+    if (job->state == FLUX_JOB_STATE_NEW) {
+        errprintf (error,
+                   "job state (%s) is invalid after replay",
+                   flux_job_statetostr (job->state, "L"));
         goto inval;
+    }
```

Comment:
Pretty unlikely since a kvs checkpoint would have to land at just the right time, but sounds like "famous last words". 

Maybe the restart code should detect some obviously wrong and unrecoverable job conditions and just remove those from the KVS and continue with startup.

I'd like to add a `version` to the submit event.  That would make it easier to distinguish some obviously wrong stuff in the eventlog from changes like the one we just had.

Let me see what I can come up with.

Reply:
Is this actually possible since the `submit` and `validate` or `invalidate` are both posted from the `submit_job()` function? That is,  I don't see a way for a job eventlog to be committed to the KVS with only a `submit` event since when `submit_job()` returns the eventlog should either have `validate` or `invalidate` as the last event (and for `invalidate` I think the eventlog is discarded?)

Especially if the broker crashes, it hasn't yet had a chance to commit the initial eventlog and respond to the submission RPC, so the job is just lost.

However, I'm not necessarily arguing against protecting against unlikely errors in eventlogs during replay. I agree with @garlick that we should move past those during startup so that these problems don't require manual intervention in the future.

Reply:
Ah good point.

I got kind of lost refactoring the restart code this morning to catch these types of cases.  Perhaps it could go in another PR so we can get this moving forward.

I'll push my changes to add a version number, which makes the backwards compatibility code less hacky.

Reply:
> Maybe the restart code should detect some obviously wrong and unrecoverable job conditions and just remove those from the KVS and continue with startup.

Why would this be a unrecoverable job condition?  Is there something about how `validate` is triggered?

Reply:
Yes, before validate is posted, the job is not officially accepted.  Only _after_ it is accepted is the user given the job ID.  So if we are restarting and the job is in NEW state, there is no way to move it forward because the user already received an error.

Reply:
With the fix proposed here, the old format is detected and state is advanced to DEPEND after submit event is processed.

Reply:
> Yeah, this was also made more likely by breaking the recursion of the event posting code, so we should definitely handle the NEW case as you suggested.

This discussion is getting confusing and I can't tell if we are all on the same page or not.  Let me summarize my understanding of how things work:
- ingest module commits jobspec not eventlog
- job manager posts a stream of events in response to submit that all land in the same batch (kvs txn) with no opportunity for it to be partially committed (recall that the recursion fix doesn't cause events to be asynchronously processed - they are just processed serially now with reentry contained in the event posting function)
- if the job is invalid, the kvs batch is committed, but upon completion, an error response is returned to ingest which unlinks the entire job directory

It helped me to walk through that :-)  Yes a version 1 job in NEW state could occur if a KVS checkpoint happened right between the job manager eventlog commit and the ingest unlink, and then a crash occurred before the next checkpoint.

I hope that makes sense!

If so, my vote would be to open an issue on that and handle it elsewhere because it will require some refactoring of the restart code that would slow down getting 0.41.1 out.

Reply:
> If so, my vote would be to open an issue on that and handle it elsewhere because it will require some refactoring of the restart code that would slow down getting 0.41.1 out.

Agreed!

Reply:
>  Yes a version 1 job in NEW state could occur if a KVS checkpoint happened right between the job manager eventlog commit and the ingest unlink

Ah, since `invalidate` doesn't transition the state out of NEW? Could the restart code just discard and unlink eventlogs for invalidated jobs?

Reply:
>  that would slow down getting 0.41.1 out.

I'm actually working on a 0.41.0-3 RPM now with patches applied. I don't think there is time to generate a new release before the OS freeze date since this PR hasn't yet landed (and we'd still need release notes, push a tag, etc.) Of course, if there is time, nothing prevents us from still releasing a 0.41.1.

Reply:
Right - the idea was that an invalid job is wiped out immediately, so no next state is defined.  In fact `invalidate` is a journal only event, simply because it seemed daft to post it right before the job directory is to be unlinked.

Ah you are right, we could close this gap simply by tacking an eventlog unlink onto the job manager's txn if the job is invalid.

Reply:
If someone wants to give the current patch a once over we could land it.  I think it's good to go as is.  I was just trying to boost the coverage a bit but it's code that wasn't covered before.